### ISSUE TITLE
Add Support For Forks

### ISSUE THREAD

**Issue by codethejason** [bitcoin-abe/bitcoin-abe#245]
Whenever trying to add a fork (i.e. Bitcoin Cash) with any similar blocks as the original chain, Abe produces an error: 
```
IntegrityError: (1062, "Duplicate entry '36572175-83525054' for key 'PRIMARY'")
```

Could Abe add support to prevent this from happening? I propose either removing the primary key restriction or allowing the error to go through since the fork has the same block hash and properties as the parent.

**Comment by dermoth** [comment]
Do you know what table is that? PK must be unique otherwise the database wouldn't function properly... in the end the block heights are also associated with a chain so multiple chains can be present in parallel.

Did you create a new chain for BCH? It would help to see the chain's class... I'm not sure yet show it should be implemented specifically but all blocks up to the fork should be taken from the Bitcoin chain if present, then at fork point some changes should be applied to signatures iirc.

Regards

**Comment by codethejason** [comment]
@dermoth You're right, I forgot that Abe did not keep track of which chain the hash belonged to. If there was another column that indicated the chain number, perhaps, it could be implemented. 

I didn't do this with BCH but I did it with another coin with a Sha256PosChain policy (blend of Sha256Chain and PpcPosChain) that ordinarily works on a few coins I've tested it on. I created a new chain for the fork and tried modifying the last block entry of the forked chain in the chain table to point to the block right before the two chains diverged, but I cannot specify a different next block because the block_next table just serves as a dictionary for the current block and the next block without taking into account the chain the block is in. I think the difference in signatures would allow Abe to tell which block was forked, but the forked block still remains an issue.

**Comment by dermoth** [comment]
Ok so I think block_next should take a chain_id as well, so the same block relationship could be used for different chains. It may affect other block_* tables though... I would have to take a serious look at it and it's one my priority right now.

If you manage to get it to work with the appropriate migrations to upgrade existing databases I'd be willing to merge your PR. Let me know if you have other questions...

Worst case you can always run Abe in different DB's/Schemas... that's what I've always did for Testnet vs Bitcoin. Besides the forks I think one advantage for a single schema was for Merged mining such as BTC/NMC... The same block would be inserted for both chains, but even then I'm not 100% sure it got implemented all the way.

**Comment by codethejason** [comment]
That's true, I'll see if I can figure it out. Otherwise, running two different databases doesn't seem to be too bad.
